### PR TITLE
Test s3 uploading parquet files from go interfaces

### PR DESCRIPTION
**Pull request by Sami1309**

### CODE REVIEW

File: provider/spark.go [hunk 1]
```diff
@@ -245,6 +283,205 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	}
+	schemaStruct := arr.Index(0)
+	typeOfS := schemaStruct.Type()
+	schemaString := `
+    {
+        "Tag":"name=parquet-go-root",
+        "Fields":[
+		    `
+	for i := 0; i < schemaStruct.NumField(); i++ {
+		fieldDataType := reflect.TypeOf(schemaStruct.Field(i).Interface()).String()
```
Comment (sdreyer) [inline provider/spark.go:343]: This can be a function

File: provider/spark.go [hunk 2]
```diff
@@ -245,6 +283,205 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	ctx := context.Background()
+	schemaString, err := s.generateSchemaFromInterface(data)
+	if err != nil {
+		return err
+	}
+	dataString, err := s.generateJSONFromInterface(data)
+	if err != nil {
+		return err
+	}
+	file, err := parquetGo.NewS3FileWriter(ctx, s.bucketPath, path, "bucket-owner-full-control", nil, &awsV1.Config{
```
Comment (sdreyer) [inline provider/spark.go:383]: I think this can be broken into
a GetWriter and Write function

File: provider/spark.go [hunk 3]
```diff
@@ -245,6 +283,205 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	num := int(pr.GetNumRows())
+	res, err := pr.ReadByNumber(num)
+	if err != nil {
+		return err
+	}
+	compareArray := reflect.ValueOf(data)
+	if len(res) != compareArray.Len() {
+		return fmt.Errorf("data do not have the same number of rows")
+	}
+	for i := 0; i < compareArray.Len(); i++ {
```
Comment (sdreyer) [inline provider/spark.go:446]: This whole block needs to be broken up

File: provider/spark.go [hunk 4]
```diff
@@ -108,7 +130,7 @@ func (store *SparkOfflineStore) AsOfflineStore() (OfflineStore, error) {
 	return store, nil
 }
 
-func sparkOfflineStoreFactory(config SerializedConfig) (Provider, error) {
+func SparkOfflineStoreFactory(config SerializedConfig) (Provider, error) {
```
[the remark this answers is not in the document]
Comment (Sami1309) [reply provider/spark.go:134]: I use it to do a local test outside of the package

File: provider/spark.go (hunk 3 above)
Comment (Sami1309) [reply provider/spark.go:510]: broke it

File: provider/spark.go [hunk 5]
```diff
@@ -5,11 +5,26 @@ import (
...
+	credentialsV1 "github.com/aws/aws-sdk-go/aws/credentials"
+	session "github.com/aws/aws-sdk-go/aws/session"
+	s3manager "github.com/aws/aws-sdk-go/service/s3/s3manager"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/credentials"
 	"github.com/aws/aws-sdk-go-v2/service/emr"
 	"github.com/aws/aws-sdk-go-v2/service/s3"
+
+	// emrTypes "github.com/aws/aws-sdk-go-v2/service/emr/types"
```
[the remark this answers is not in the document]
Comment (Sami1309) [reply provider/spark.go:23]: removed

File: provider/tests/scripts/offline_store_spark_runner.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import argparse
```
Comment (Sami1309) [inline provider/tests/scripts/offline_store_spark_runner.py:1]: this file needs to exist to initialize a spark store

File: provider/spark.go [hunk 7]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+		Region:      awsV1.String(s.region),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return fr, nil
+}
+
+func typeConvertValue(value interface{}) interface{} {
+	switch reflect.TypeOf(value).String() {
```
Comment (sdreyer) [inline provider/spark.go:469]: Does this need a default value?

File: provider/spark.go [hunk 8]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	case "time.Time":
+		return value.(time.Time).UnixMilli()
+	case "int":
+		return int32(value.(int))
+	}
+	return value
+}
+
+func compareStructs(local interface{}, fetched interface{}) error {
```
Comment (sdreyer) [inline provider/spark.go:478]: Isnt there built in functions to compare structs?

File: provider/spark.go [hunk 9]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
 	return nil, nil
 }
 
+func (s *S3Store) ResourcePath(id ResourceID) string {
+	return fmt.Sprintf("s3://%s/%s", s.bucketPath, s.resourcePrefix(id))
+}
+
+func stringifyValue(value interface{}) string {
```
Comment (sdreyer) [inline provider/spark.go:290]: Unit test

File: provider/spark.go [hunk 10]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
 	return nil, nil
 }
 
+func (s *S3Store) ResourcePath(id ResourceID) string {
```
Comment (sdreyer) [inline provider/spark.go:286]: unit test

File: provider/spark.go [hunk 11]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	case "bool":
+		return fmt.Sprintf(`%t`, value.(bool))
+	case "time.Time":
+		//convert to int64
+		return fmt.Sprintf(`%d`, value.(time.Time).UnixMilli())
+	}
+	return ""
+}
+
+func stringifyStruct(data interface{}) string {
```
Comment (sdreyer) [inline provider/spark.go:311]: unit test

File: provider/spark.go [hunk 12]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+		if j != curStruct.NumField()-1 {
+			structString += ",\n"
+		} else {
+			structString += "\n}"
+		}
+	}
+	return structString
+}
+
+func stringifyStructArray(data interface{}) ([]string, error) {
```
Comment (sdreyer) [inline provider/spark.go:329]: unit test

File: provider/spark.go [hunk 13]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	array := reflect.ValueOf(data)
+	structStringArray := make([]string, array.Len())
+	for i := 0; i < array.Len(); i++ {
+		structStringArray[i] = stringifyStruct(array.Index(i).Interface())
+	}
+	return structStringArray, nil
+}
+
+func stringifyStructField(data interface{}, idx int) string {
```
Comment (sdreyer) [inline provider/spark.go:338]: unit test

File: provider/spark.go [hunk 14]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+		return fmt.Sprintf(`{"Tag": "name=%s, type=INT32"}`, jsonFriendlyFieldName)
+	case "bool":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=BOOLEAN"}`, jsonFriendlyFieldName)
+	case "time.Time":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=INT64"}`, jsonFriendlyFieldName)
+	}
+	return ""
+}
+
+func parquetSchemaHeader() string {
```
Comment (sdreyer) [inline provider/spark.go:362]: unit test

File: provider/spark.go [hunk 15]
```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
...
+	return `
+    {
+        "Tag":"name=parquet-go-root",
+        "Fields":[
+		    `
+}
+
+func generateSchemaFromInterface(data interface{}) (string, error) {
```
Comment (sdreyer) [inline provider/spark.go:370]: unit test

File: provider/spark.go (hunk 14 above)
Comment (Sami1309) [reply provider/spark.go:362]: should I just make it a variable since it's just a single string return?

Comment (sdreyer) [reply provider/spark.go:362]: Yeah that's fine

**Review by sdreyer — APPROVED** [review]

File: provider/spark.go [hunk 16]
```diff
@@ -155,16 +184,18 @@ type EMRExecutor struct {
...
+	region      string
+	bucketPath  string
 }
 
 func (s *S3Store) UploadSparkScript() error {
 	var sparkScriptPath string
 	sparkScriptPath, ok := os.LookupEnv("SPARK_SCRIPT_PATH")
 	if !ok {
-		sparkScriptPath = "./scripts/offline_store_spark_runner.py"
+		sparkScriptPath = "./provider/scripts/offline_store_spark_runner.py"
```
Comment (sdreyer) [inline provider/spark.go:198]: Can we remove the provider/provider/scripts directory nesting?

**Review by sdreyer — APPROVED** [review]

**Pull request merged**